### PR TITLE
Kv store grpc

### PR DESCRIPTION
**Pull request by floan**
Added the GRPC functionality to my KVStore class (Client and Server Func)

### CODE REVIEW

**Comment by floan** [comment]
@adamegyed @barath was wondering if this is fine? Also, can we send pull requests over the weekend?

File: src/key_value_store/server/kvstore_server_runner.cpp [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+  // clients. In this case it corresponds to an *synchronous* service.
+  builder.RegisterService(&service);
+  // Finally assemble the server.
+  std::unique_ptr<Server> server(builder.BuildAndStart());
+  std::cout << "Server listening on " << server_address << std::endl;
+
+  // Wait for the server to shutdown. Note that some other thread must be
+  // responsible for shutting down the server for this call to ever return.
+  server->Wait();
+
```
Comment (barath) [inline src/key_value_store/server/kvstore_server_runner.cpp]: Remove blank line

File: src/key_value_store/server/kvstore_server.h [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+using kvstore::PutReply;
+using kvstore::GetRequest;
+using kvstore::GetReply;
+using kvstore::RemoveRequest;
+using kvstore::RemoveReply;
+
+class KeyValueStoreImpl final :
+public kvstore::KeyValueStore::Service {
+ public:
+
```
Comment (barath) [inline src/key_value_store/server/kvstore_server.h]: Remove blank line

File: src/key_value_store/server/kvstore_server.h [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+using grpc::StatusCode;
+
+using kvstore::PutRequest;
+using kvstore::PutReply;
+using kvstore::GetRequest;
+using kvstore::GetReply;
+using kvstore::RemoveRequest;
+using kvstore::RemoveReply;
+
+class KeyValueStoreImpl final :
```
Comment (barath) [inline src/key_value_store/server/kvstore_server.h]: Put on one line.  (You may want to run clang-format.)

File: src/key_value_store/server/kvstore_server.cpp [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+  }
+}
+
+Status KeyValueStoreImpl::get(ServerContext* context,
+    ServerReaderWriter<GetReply, GetRequest>* stream) {
+  GetRequest request;
+  while (stream->Read(&request)) {
+    // Get the value vector and store it in values
+    std::vector<std::string> values(store_.Get(request.key()));
+    for (std::string s : values) {
```
Comment (barath) [inline src/key_value_store/server/kvstore_server.cpp]: const reference is better

File: src/key_value_store/server/kvstore_server.cpp [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+      stream->Write(response);
+    }
+  }
+  return Status::OK;
+}
+
+Status KeyValueStoreImpl::remove(ServerContext* context,
+    const RemoveRequest* request, RemoveReply* response) {
+  bool result = store_.Remove(request->key());
+  if (result) {
```
Comment (barath) [inline src/key_value_store/server/kvstore_server.cpp]: Perhaps select a better variable name

File: src/key_value_store/server/kvstore_server.cpp [hunk 6]
```diff
@@ -0,0 +1,38 @@
+#include "kvstore_server.h"
+
+#include <vector>
+
+Status KeyValueStoreImpl::put(ServerContext* context, 
+    const PutRequest* request, PutReply* response) {
+  bool result = store_.Put(request->key(), request->value());
```
Comment (barath) [inline src/key_value_store/server/kvstore_server.cpp]: Select a better variable name

File: src/key_value_store/client/kvstore_client.h [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+using grpc::Status;
+
+using kvstore::PutRequest;
+using kvstore::PutReply;
+using kvstore::GetRequest;
+using kvstore::GetReply;
+using kvstore::RemoveRequest;
+using kvstore::RemoveReply;
+
+class KeyValueStoreClient final : public KeyValueStoreInterface {
```
Comment (barath) [inline src/key_value_store/client/kvstore_client.h:26]: Add a comment for the class, all member methods, and member vars.  (Here and elsewhere.)

File: src/key_value_store/client/kvstore_client.h [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+ public:
+
```
Comment (barath) [inline src/key_value_store/client/kvstore_client.h]: Remove blank line.

File: src/key_value_store/client/kvstore_client.cpp [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+  RemoveReply response; 
+  ClientContext context;
+  request.set_key(key);
+
+  Status status = stub_->put(&context, request, &response);
+
+  if (status.ok()) {
+    return true;
+  } else {
+    std::cout << status.error_code() << ": " << status.error_message()
```
Comment (barath) [inline src/key_value_store/client/kvstore_client.cpp]: Better to use glog than cout

Comment (floan) [reply src/key_value_store/client/kvstore_client.cpp]: Made this change.

File: src/key_value_store/server/kvstore_server.cpp (hunk 6 above)
Comment (floan) [reply src/key_value_store/server/kvstore_server.cpp]: changed

File: src/key_value_store/server/kvstore_server.cpp (hunk 4 above)
Comment (floan) [reply src/key_value_store/server/kvstore_server.cpp]: changed

File: src/key_value_store/server/kvstore_server.cpp (hunk 5 above)
Comment (floan) [reply src/key_value_store/server/kvstore_server.cpp]: changed

File: src/key_value_store/server/kvstore_server.h (hunk 2 above)
Comment (floan) [reply src/key_value_store/server/kvstore_server.h]: done

File: src/key_value_store/server/kvstore_server.h (hunk 3 above)
Comment (floan) [reply src/key_value_store/server/kvstore_server.h]: done, ran this on all files

File: src/key_value_store/server/kvstore_server_runner.cpp (hunk 1 above)
Comment (floan) [reply src/key_value_store/server/kvstore_server_runner.cpp]: done

File: src/key_value_store/client/kvstore_client.h (hunk 7 above)
Comment (floan) [reply src/key_value_store/client/kvstore_client.h:26]: done

File: src/key_value_store/client/kvstore_client.h (hunk 8 above)
Comment (floan) [reply src/key_value_store/client/kvstore_client.h]: done

**Review by floan — COMMENTED** [review]
Have made requested changes. Many edits were made by clang-format, that is why you might see numerous additions / deletions.

File: src/key_value_store/KeyValueStoreInterface.h [hunk 10]
```diff
@@ -0,0 +1,11 @@
+#include <string>
+#include <vector>
+
+class KeyValueStoreInterface {
```
Comment (adamegyed) [inline src/key_value_store/KeyValueStoreInterface.h:9]: Add a comment explaining why this interface exists

File: src/key_value_store/server/kvstore_server_runner.cpp [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+  // Wait for the server to shutdown. Note that some other thread must be
+  // responsible for shutting down the server for this call to ever return.
+  server->Wait();
+}
+
+int main(int argc, char **argv) {
+  RunServer();
+
+  return 0;
+}
```
Comment (adamegyed) [inline src/key_value_store/server/kvstore_server_runner.cpp]: Merge this into the same executable build target as the other, empty main in the kvstore folder

File: src/key_value_store/client/kvstore_client.h [hunk 12]
```diff
@@ -1,35 +1,41 @@
...
+// grpc and process the result to cpp syntax output
 class KeyValueStoreClient final : public KeyValueStoreInterface {
- public:
-
+public:
   KeyValueStoreClient(std::shared_ptr<Channel> channel)
-    : stub_(Greeter::NewStub(channel)) {};
-
+      : stub_(Greeter::NewStub(channel)){};
+  // Performs normal put functionality
```
Comment (barath) [inline src/key_value_store/client/kvstore_client.h:32]: Arguments should be const references.

Also, comments should be more clear / detailed, explaining the arguments and return value.

File: src/key_value_store/KeyValueStoreInterface.h (hunk 10 above)
Comment (floan) [reply src/key_value_store/KeyValueStoreInterface.h:9]: Done

File: src/key_value_store/client/kvstore_client.h (hunk 12 above)
Comment (floan) [reply src/key_value_store/client/kvstore_client.h:32]: Have expanded on them. Please let me know if I should elaborate further.

File: src/key_value_store/server/kvstore_server_runner.cpp (hunk 11 above)
Comment (floan) [reply src/key_value_store/server/kvstore_server_runner.cpp]: Done!

**Review by floan — COMMENTED** [review]
Have made the changes. Please review whenever possible.

File: src/key_value_store/client/kvstore_client.cpp [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+
+bool KeyValueStoreClient::Put(const std::string &key,
+                              const std::string &value) {
+  PutRequest request;
+  PutReply response;
+  ClientContext context;
+  request.set_key(key);
+  request.set_value(value);
+
+  // TODO: ASK WHY THE REQ OBJ IS SENT INSTEAD OF PTR
```
Comment (floan) [inline src/key_value_store/client/kvstore_client.cpp:15]: Also @barath  @adamegyed  I was really confused about this. Even though the type in the function is a PutRequest*, we are passing in a request object. Intuitively I'd say this is wrong but this is how Google did it in their example so...

**Review by barath — APPROVED** [review]

**Pull request merged**